src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDoclet.java

```diff
@@ -169,18 +169,20 @@ protected void generateOtherFiles(DocletEnvironment docEnv, ClassTree classtree)
         }
 
         if (options.createIndex()) {
-            IndexBuilder indexBuilder = new IndexBuilder(configuration, nodeprecated);
+            configuration.mainIndex.addElements();
             if (options.splitIndex()) {
-                SplitIndexWriter.generate(configuration, indexBuilder);
+                SplitIndexWriter.generate(configuration);
             } else {
-                SingleIndexWriter.generate(configuration, indexBuilder);
+                SingleIndexWriter.generate(configuration);
             }
-            AllClassesIndexWriter.generate(configuration,
-                    new IndexBuilder(configuration, nodeprecated, true));
+            IndexBuilder allClassesIndex = new IndexBuilder(configuration, nodeprecated, true);
+            allClassesIndex.addElements();
+            AllClassesIndexWriter.generate(configuration, allClassesIndex);
             if (!configuration.packages.isEmpty()) {
                 AllPackagesIndexWriter.generate(configuration);
             }
             SystemPropertiesWriter.generate(configuration);
```

Comment:
Logically, it would make more sense to write this file before writing the index files, in case the index files depend in any way on any for the individual pages being written.